File: docs/site/samples/python/examples/guides_derive_render_variant.py
"""Derive guide — render a named variant: after adding a variant with ``derive``, select
it by name at render time. Variant selection is caller-owned.

Standalone — the docs page displays this file verbatim; run it directly to check.
"""

from pathlib import Path

from prompting_press import Prompt
from pydantic import BaseModel, field_validator

# The caller reads the definition; the library does no file I/O itself.
# Resolve the file next to this program (a real app uses its own path).
_HERE = Path(__file__).parent


class AssistantVars(BaseModel):
    company: str
    max_words: int

    @field_validator("max_words")
    @classmethod
    def _at_least_one(cls, v: int) -> int:
        if v < 1:
            raise ValueError("max_words must be at least 1")
        return v


def main() -> None:
    assistant = Prompt.from_yaml((_HERE / "assistant.yaml").read_text())
    derived = assistant.derive(
        {
            "variants": {
                **assistant.variants,
                "formal": {
                    "body": "You are the official support assistant for {{ company }}. "
                    "Please keep every reply under {{ max_words }} words."
                },
            }
        }
    )

    result = derived.render(
        AssistantVars,
        data={"company": "Acme Robotics", "max_words": 50},
        variant="formal",
    )
    print(
        result.text
    )  # "You are the official support assistant for Acme Robotics. ..."
    print(result.variant)  # "formal"
    assert (
        result.text == "You are the official support assistant for Acme Robotics. "
        "Please keep every reply under 50 words."
    )
    assert result.variant == "formal"


if __name__ == "__main__":
    main()

File: docs/site/samples/typescript/examples/guides_derive_render_variant.ts
/**
 * Derive guide — render a named variant: after adding a variant with `derive`, select it by
 * name at render time. Variant selection is caller-owned.
 *
 * Standalone — the docs page displays this file verbatim; run it directly to check.
 */

import assert from "node:assert/strict";
import { test } from "node:test";
import { Prompt } from "prompting-press";
import { z } from "zod";

const assistantYaml = `
name: assistant
role: system
body: "You are a support assistant for {{ company }}. Keep your replies under {{ max_words }} words."
variables:
  company: { type: string, trusted: true }
  max_words: { type: integer, trusted: true }
`;

const AssistantVars = z.object({
	company: z.string().min(1),
	max_words: z.number().int().min(1),
});

test("a derived named variant renders by name", () => {
	const assistant = Prompt.fromYaml(assistantYaml);
	const derivedAssistant = assistant.derive({
		variants: {
			...assistant.variants,
			formal: {
				body: "You are the official support assistant for {{ company }}. Please keep every reply under {{ max_words }} words.",
			},
		},
	});

	const result = derivedAssistant.render(
		AssistantVars,
		{ company: "Acme Robotics", max_words: 50 },
		{ variant: "formal" },
	);
	console.log(result.text); // "You are the official support assistant for Acme Robotics. Please keep every reply under 50 words."
	console.log(result.variant); // "formal"
	assert.equal(
		result.text,
		"You are the official support assistant for Acme Robotics. Please keep every reply under 50 words.",
	);
	assert.equal(result.variant, "formal");
});
